### PR TITLE
[query/pcode] make gqFromPL use registerPCode

### PR DESCRIPTION
**Pull request by chrisvittal**

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/functions/GenotypeFunctions.scala [hunk 1]
```diff
@@ -10,32 +10,32 @@ import is.hail.expr.types.virtual.{TArray, TFloat64, TInt32, Type}
...
-              m2 := pli,
-              Code._empty)),
-          i := i + 1
-        ),
-        m2 - m)
+    registerPCode("gqFromPL", TArray(tv("N", "int32")), TInt32, (_: Type, _: PType) => PInt32())
+    { case (r, rt, _pl: PIndexableCode) =>
+      val code = EmitCodeBuilder.scopedCode(r.mb) { cb =>
+        val pl = _pl.memoize(cb, "plv")
+        val m = cb.memoize[Int](const(99), "m")
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/ir/functions/GenotypeFunctions.scala:17]: not specific to this PR, but why is this called memoize?

File: hail/src/main/scala/is/hail/expr/ir/functions/GenotypeFunctions.scala [hunk 2]
```diff
@@ -10,32 +10,32 @@ import is.hail.expr.types.virtual.{TArray, TFloat64, TInt32, Type}
...
+        val m2 = cb.memoize[Int](const(99), "m2")
+        val i = cb.memoize[Int](const(0), "i")
+
+        cb.whileLoop(i < pl.loadLength(), {
+          val iec = pl.loadElement(cb, i)
+          cb += iec.Lmissing
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/ir/functions/GenotypeFunctions.scala:23]: append or += throughout

Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/functions/GenotypeFunctions.scala:23]: not sure why define exists, but that could be used too

File: hail/src/main/scala/is/hail/expr/ir/functions/GenotypeFunctions.scala (hunk 1 above)
Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/functions/GenotypeFunctions.scala:17]: I see from the perspective that we don't just re-run the first argument, but store it instead. That's a bit different than what I think of as memoization (checking if a value is already stored, say in a hash table by looking up some key), our memoization is just an assignment.

File: hail/src/main/scala/is/hail/expr/ir/functions/GenotypeFunctions.scala [hunk 3]
```diff
@@ -10,32 +10,32 @@ import is.hail.expr.types.virtual.{TArray, TFloat64, TInt32, Type}
...
+          cb += Code._fatal[Unit]("PL cannot have missing elements.")
+          cb += iec.Lpresent
+          val pli = cb.memoize[Int](iec.pc.tcode[Int], "pli")
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/ir/functions/GenotypeFunctions.scala:26]: I would prefer changes like this (.tcode[Int] away from Region.loadInt) to be made in a separate PR, out of scope from the move to registerPCode.

Comment (chrisvittal) [reply hail/src/main/scala/is/hail/expr/ir/functions/GenotypeFunctions.scala:26]: That is the scope of moving to registerPCode.

This change is two things:
1. Replacement of the single address `pl` variable with a `PIndexableValue` (created by `PIndexableCode.memoize`
1. Now that we have a `PIndexableValue`, use `loadElement` which returns an `IEmitCode` which encapsulates all of `isElementDefined`, `loadElement` (which only gets you the address), and `loadInt`.

Everything else is necessary for these two changes to take place, but not fundamental to what this change is.

**Comment by chrisvittal** [comment]
The only thing that may be a good change is to add a more descriptive name for variable creation and initialization than `memoize`. Because you're right, it doesn't really memoize. That's outside the scope of this PR however.

File: hail/src/main/scala/is/hail/expr/ir/functions/GenotypeFunctions.scala (hunk 2 above)
Comment (akotlar) [reply hail/src/main/scala/is/hail/expr/ir/functions/GenotypeFunctions.scala:23]: You do however use "assign" in "user" code (which is append(local := value)). It would be nice to settle on methods or infix operators. I do agree that having a visual aid to make generated code look more like interpreted code, although portions like this make that aid look inconsistent:

```scala
cb += iec.Lmissing
cb += Code._fatal[Unit]("PL cannot have missing elements.")
cb += iec.Lpresent
```
which looks very different than an if statement.

But if you feel strongly this is the right mix of method / infix, no problem.
